File: repertoire/data-integration/providers/enricher_plugin/exif_extract.ts
// COPF Data Integration Kit - EXIF metadata extraction enricher provider
// Reads EXIF IFD entries from JPEG/TIFF bytes: parses APP1 marker, reads IFD tags for
// camera, lens, GPS, datetime, dimensions.

export const PROVIDER_ID = 'exif_extract';
export const PLUGIN_TYPE = 'enricher_plugin';

export interface ContentItem {
  id: string;
  content: string;
  contentType: string;
  metadata?: Record<string, unknown>;
}

export interface EnricherConfig {
  model?: string;
  apiKey?: string;
  threshold?: number;
  options?: Record<string, unknown>;
}

export interface EnrichmentResult {
  fields: Record<string, unknown>;
  confidence: number;
  metadata?: Record<string, unknown>;
}

export interface SchemaRef {
  name: string;
  fields?: string[];
}

export interface CostEstimate {
  tokens?: number;
  apiCalls?: number;
  durationMs?: number;
}

// EXIF IFD tag constants
const EXIF_TAGS: Record<number, string> = {
  0x010F: 'Make',             // Camera manufacturer
  0x0110: 'Model',            // Camera model
  0x0112: 'Orientation',      // Image orientation
  0x011A: 'XResolution',      // Horizontal resolution
  0x011B: 'YResolution',      // Vertical resolution
  0x0132: 'DateTime',         // File change date
  0x8769: 'ExifIFDPointer',   // Pointer to Exif sub-IFD
  0x8825: 'GPSInfoIFDPointer', // Pointer to GPS IFD
  0x9003: 'DateTimeOriginal', // Original datetime
  0x9004: 'DateTimeDigitized',
  0x920A: 'FocalLength',
  0xA405: 'FocalLengthIn35mmFilm',
  0x829A: 'ExposureTime',
  0x829D: 'FNumber',
  0x8827: 'ISOSpeedRatings',
  0xA001: 'ColorSpace',
  0xA002: 'ExifImageWidth',
  0xA003: 'ExifImageHeight',
  0xA434: 'LensModel',
  0xA433: 'LensMake',
};

const GPS_TAGS: Record<number, string> = {
  0x0001: 'GPSLatitudeRef',
  0x0002: 'GPSLatitude',
  0x0003: 'GPSLongitudeRef',
  0x0004: 'GPSLongitude',
  0x0005: 'GPSAltitudeRef',
  0x0006: 'GPSAltitude',
};

class ExifReader {
  private buffer: Buffer;
  private littleEndian: boolean = false;
  private tiffOffset: number = 0;

  constructor(buffer: Buffer) {
    this.buffer = buffer;
  }

  readUint16(offset: number): number {
    return this.littleEndian
      ? this.buffer.readUInt16LE(offset)
      : this.buffer.readUInt16BE(offset);
  }

  readUint32(offset: number): number {
    return this.littleEndian
      ? this.buffer.readUInt32LE(offset)
      : this.buffer.readUInt32BE(offset);
  }

  readRational(offset: number): number {
    const numerator = this.readUint32(offset);
    const denominator = this.readUint32(offset + 4);
    return denominator !== 0 ? numerator / denominator : 0;
  }

  readString(offset: number, length: number): string {
    let result = '';
    for (let i = 0; i < length; i++) {
      const code = this.buffer[offset + i];
      if (code === 0) break;
      result += String.fromCharCode(code);
    }
    return result;
  }

  parse(): Record<string, unknown> {
    const result: Record<string, unknown> = {};

    // Find JPEG APP1 marker (0xFFE1) containing EXIF data
    let offset = 0;
    if (this.buffer[0] === 0xFF && this.buffer[1] === 0xD8) {
      // JPEG file - scan for APP1 marker
      offset = 2;
      while (offset < this.buffer.length - 1) {
        if (this.buffer[offset] === 0xFF && this.buffer[offset + 1] === 0xE1) {
          offset += 4; // Skip marker + length
          break;
        }
        if (this.buffer[offset] === 0xFF) {
          const segmentLength = this.buffer.readUInt16BE(offset + 2);
          offset += 2 + segmentLength;
        } else {
          offset++;
        }
      }
    }

    // Verify "Exif\0\0" header
    const exifHeader = this.readString(offset, 4);
    if (exifHeader !== 'Exif') return result;
    offset += 6; // Skip "Exif\0\0"

    this.tiffOffset = offset;

    // Read TIFF header to determine byte order
    const byteOrder = this.buffer.readUInt16BE(offset);
    this.littleEndian = byteOrder === 0x4949; // 'II' = Intel = little-endian

    // Verify TIFF magic number (42)
    const magic = this.readUint16(offset + 2);
    if (magic !== 42) return result;

    // Read offset to first IFD
    const ifd0Offset = this.readUint32(offset + 4);

    // Parse IFD0 (main image IFD)
    const ifd0Tags = this.parseIFD(this.tiffOffset + ifd0Offset);

    for (const [tag, value] of ifd0Tags) {
      const tagName = EXIF_TAGS[tag];
      if (tagName) {
        result[tagName] = value;
      }
    }

    // Follow ExifIFDPointer to sub-IFD
    const exifIfdPtr = ifd0Tags.get(0x8769);
    if (typeof exifIfdPtr === 'number') {
      const exifTags = this.parseIFD(this.tiffOffset + exifIfdPtr);
      for (const [tag, value] of exifTags) {
        const tagName = EXIF_TAGS[tag];
        if (tagName) result[tagName] = value;
      }
    }

    // Follow GPSInfoIFDPointer
    const gpsIfdPtr = ifd0Tags.get(0x8825);
    if (typeof gpsIfdPtr === 'number') {
      const gpsTags = this.parseIFD(this.tiffOffset + gpsIfdPtr);
      const gps: Record<string, unknown> = {};
      for (const [tag, value] of gpsTags) {
        const tagName = GPS_TAGS[tag];
        if (tagName) gps[tagName] = value;
      }

      // Convert GPS coordinates to decimal degrees
      if (gps.GPSLatitude && gps.GPSLongitude) {
        const lat = this.gpsToDecimal(gps.GPSLatitude as number[], gps.GPSLatitudeRef as string);
        const lon = this.gpsToDecimal(gps.GPSLongitude as number[], gps.GPSLongitudeRef as string);
        result['GPSLatitude'] = lat;
        result['GPSLongitude'] = lon;
        if (gps.GPSAltitude) result['GPSAltitude'] = gps.GPSAltitude;
      }
    }

    return result;
  }

  private parseIFD(offset: number): Map<number, unknown> {
    const tags = new Map<number, unknown>();
    if (offset >= this.buffer.length - 2) return tags;

    const entryCount = this.readUint16(offset);
    let pos = offset + 2;

    for (let i = 0; i < entryCount; i++) {
      if (pos + 12 > this.buffer.length) break;

      const tag = this.readUint16(pos);
      const type = this.readUint16(pos + 2);
      const count = this.readUint32(pos + 4);
      const valueOffset = pos + 8;

      const value = this.readTagValue(type, count, valueOffset);
      tags.set(tag, value);

      pos += 12;
    }

    return tags;
  }

  private readTagValue(type: number, count: number, valueOffset: number): unknown {
    const totalBytes = this.typeSize(type) * count;
    const dataOffset = totalBytes > 4
      ? this.tiffOffset + this.readUint32(valueOffset)
      : valueOffset;

    switch (type) {
      case 1: // BYTE
        return count === 1 ? this.buffer[dataOffset] : Array.from(this.buffer.slice(dataOffset, dataOffset + count));
      case 2: // ASCII
        return this.readString(dataOffset, count);
      case 3: // SHORT
        return count === 1 ? this.readUint16(dataOffset) : Array.from({ length: count }, (_, i) => this.readUint16(dataOffset + i * 2));
      case 4: // LONG
        return count === 1 ? this.readUint32(dataOffset) : Array.from({ length: count }, (_, i) => this.readUint32(dataOffset + i * 4));
      case 5: // RATIONAL
        if (count === 1) return this.readRational(dataOffset);
        return Array.from({ length: count }, (_, i) => this.readRational(dataOffset + i * 8));
      default:
        return this.readUint32(valueOffset);
    }
  }

  private typeSize(type: number): number {
    switch (type) {
      case 1: case 2: return 1;
      case 3: return 2;
      case 4: return 4;
      case 5: return 8;
      default: return 1;
    }
  }

  private gpsToDecimal(dms: number[], ref: string): number {
    if (dms.length < 3) return 0;
    let decimal = dms[0] + dms[1] / 60 + dms[2] / 3600;
    if (ref === 'S' || ref === 'W') decimal = -decimal;
    return Math.round(decimal * 1000000) / 1000000;
  }
}

export class ExifExtractEnricherProvider {
  async enrich(item: ContentItem, config: EnricherConfig): Promise<EnrichmentResult> {
    const imageBuffer = Buffer.from(item.content, 'base64');
    const reader = new ExifReader(imageBuffer);
    const exifData = reader.parse();

    const hasData = Object.keys(exifData).length > 0;

    // Structure the output into logical groups
    const structured: Record<string, unknown> = {
      camera: {
        make: exifData.Make ?? null,
        model: exifData.Model ?? null,
        lens: exifData.LensModel ?? null,
        lensMake: exifData.LensMake ?? null,
      },
      settings: {
        exposureTime: exifData.ExposureTime ?? null,
        fNumber: exifData.FNumber ?? null,
        iso: exifData.ISOSpeedRatings ?? null,
        focalLength: exifData.FocalLength ?? null,
        focalLength35mm: exifData.FocalLengthIn35mmFilm ?? null,
      },
      datetime: {
        original: exifData.DateTimeOriginal ?? null,
        digitized: exifData.DateTimeDigitized ?? null,
        modified: exifData.DateTime ?? null,
      },
      dimensions: {
        width: exifData.ExifImageWidth ?? null,
        height: exifData.ExifImageHeight ?? null,
        orientation: exifData.Orientation ?? null,
      },
      gps: exifData.GPSLatitude != null ? {
        latitude: exifData.GPSLatitude,
        longitude: exifData.GPSLongitude,
        altitude: exifData.GPSAltitude ?? null,
      } : null,
    };

    return {
      fields: {
        exif: structured,
        raw_tags: exifData,
        has_gps: exifData.GPSLatitude != null,
        tag_count: Object.keys(exifData).length,
      },
      confidence: hasData ? 0.95 : 0.1,
      metadata: {
        provider: PROVIDER_ID,
        parsedTags: Object.keys(exifData).length,
        method: 'binary_ifd_parse',
      },
    };
  }

  appliesTo(schema: SchemaRef): boolean {
    const imageSchemas = ['image', 'photo', 'jpeg', 'jpg', 'tiff', 'picture'];
    return imageSchemas.some((s) => schema.name.toLowerCase().includes(s));
  }

  costEstimate(item: ContentItem): CostEstimate {
    // EXIF parsing is fast - only reads headers, not pixel data
    return { durationMs: 5, apiCalls: 0 };
  }
}

export default ExifExtractEnricherProvider;

File: repertoire/data-integration/providers/enricher_plugin/exif_extract.rs
// Clef Data Integration Kit - EXIF metadata extraction enricher provider
// Reads EXIF IFD entries from JPEG/TIFF bytes: parses APP1 marker, reads IFD tags
// for camera (0x010F), lens, GPS (IFD 0x8825), datetime (0x9003), dimensions.

use std::collections::HashMap;

pub const PROVIDER_ID: &str = "exif_extract";
pub const PLUGIN_TYPE: &str = "enricher_plugin";

#[derive(Debug, Clone)]
pub struct ContentItem {
    pub id: String,
    pub content: String,
    pub content_type: String,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct EnricherConfig {
    pub model: Option<String>,
    pub api_key: Option<String>,
    pub threshold: Option<f64>,
    pub options: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct EnrichmentResult {
    pub fields: HashMap<String, serde_json::Value>,
    pub confidence: f64,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct SchemaRef {
    pub name: String,
    pub fields: Option<Vec<String>>,
}

#[derive(Debug, Clone)]
pub struct CostEstimate {
    pub tokens: Option<u64>,
    pub api_calls: Option<u64>,
    pub duration_ms: Option<u64>,
}

#[derive(Debug)]
pub enum EnricherError {
    ParseError(String),
}

fn tag_name(tag: u16) -> Option<&'static str> {
    match tag {
        0x010F => Some("Make"),
        0x0110 => Some("Model"),
        0x0112 => Some("Orientation"),
        0x011A => Some("XResolution"),
        0x011B => Some("YResolution"),
        0x0132 => Some("DateTime"),
        0x8769 => Some("ExifIFDPointer"),
        0x8825 => Some("GPSInfoIFDPointer"),
        0x9003 => Some("DateTimeOriginal"),
        0x9004 => Some("DateTimeDigitized"),
        0x920A => Some("FocalLength"),
        0xA405 => Some("FocalLengthIn35mmFilm"),
        0x829A => Some("ExposureTime"),
        0x829D => Some("FNumber"),
        0x8827 => Some("ISOSpeedRatings"),
        0xA002 => Some("ExifImageWidth"),
        0xA003 => Some("ExifImageHeight"),
        0xA434 => Some("LensModel"),
        0xA433 => Some("LensMake"),
        _ => None,
    }
}

fn gps_tag_name(tag: u16) -> Option<&'static str> {
    match tag {
        0x0001 => Some("GPSLatitudeRef"),
        0x0002 => Some("GPSLatitude"),
        0x0003 => Some("GPSLongitudeRef"),
        0x0004 => Some("GPSLongitude"),
        0x0005 => Some("GPSAltitudeRef"),
        0x0006 => Some("GPSAltitude"),
        _ => None,
    }
}

struct ExifReader {
    data: Vec<u8>,
    little_endian: bool,
    tiff_offset: usize,
}

impl ExifReader {
    fn new(data: Vec<u8>) -> Self {
        Self { data, little_endian: false, tiff_offset: 0 }
    }

    fn read_u16(&self, offset: usize) -> u16 {
        if offset + 1 >= self.data.len() { return 0; }
        if self.little_endian {
            u16::from_le_bytes([self.data[offset], self.data[offset + 1]])
        } else {
            u16::from_be_bytes([self.data[offset], self.data[offset + 1]])
        }
    }

    fn read_u32(&self, offset: usize) -> u32 {
        if offset + 3 >= self.data.len() { return 0; }
        let bytes = [self.data[offset], self.data[offset+1], self.data[offset+2], self.data[offset+3]];
        if self.little_endian { u32::from_le_bytes(bytes) } else { u32::from_be_bytes(bytes) }
    }

    fn read_rational(&self, offset: usize) -> f64 {
        let num = self.read_u32(offset) as f64;
        let den = self.read_u32(offset + 4) as f64;
        if den != 0.0 { num / den } else { 0.0 }
    }

    fn read_string(&self, offset: usize, length: usize) -> String {
        let end = (offset + length).min(self.data.len());
        let bytes = &self.data[offset..end];
        let null_pos = bytes.iter().position(|&b| b == 0).unwrap_or(bytes.len());
        String::from_utf8_lossy(&bytes[..null_pos]).to_string()
    }

    fn parse(&mut self) -> HashMap<String, serde_json::Value> {
        let mut result = HashMap::new();
        let mut offset = 0usize;

        // Find JPEG APP1 marker
        if self.data.len() > 2 && self.data[0] == 0xFF && self.data[1] == 0xD8 {
            offset = 2;
            while offset < self.data.len().saturating_sub(1) {
                if self.data[offset] == 0xFF && self.data[offset + 1] == 0xE1 {
                    offset += 4; // Skip marker + length
                    break;
                }
                if self.data[offset] == 0xFF && offset + 3 < self.data.len() {
                    let seg_len = u16::from_be_bytes([self.data[offset+2], self.data[offset+3]]) as usize;
                    offset += 2 + seg_len;
                } else {
                    offset += 1;
                }
            }
        }

        // Verify "Exif\0\0" header
        if offset + 6 > self.data.len() { return result; }
        let header = self.read_string(offset, 4);
        if header != "Exif" { return result; }
        offset += 6;

        self.tiff_offset = offset;

        // Read byte order
        if offset + 1 >= self.data.len() { return result; }
        let byte_order = u16::from_be_bytes([self.data[offset], self.data[offset + 1]]);
        self.little_endian = byte_order == 0x4949;

        // Verify TIFF magic (42)
        let magic = self.read_u16(offset + 2);
        if magic != 42 { return result; }

        // Read first IFD offset
        let ifd0_offset = self.read_u32(offset + 4) as usize;
        let ifd0_tags = self.parse_ifd(self.tiff_offset + ifd0_offset);

        let mut exif_ifd_ptr: Option<usize> = None;
        let mut gps_ifd_ptr: Option<usize> = None;

        for (tag, value) in &ifd0_tags {
            if *tag == 0x8769 {
                if let serde_json::Value::Number(n) = value {
                    exif_ifd_ptr = n.as_u64().map(|v| v as usize);
                }
            } else if *tag == 0x8825 {
                if let serde_json::Value::Number(n) = value {
                    gps_ifd_ptr = n.as_u64().map(|v| v as usize);
                }
            }
            if let Some(name) = tag_name(*tag) {
                result.insert(name.to_string(), value.clone());
            }
        }

        // Parse Exif sub-IFD
        if let Some(ptr) = exif_ifd_ptr {
            let exif_tags = self.parse_ifd(self.tiff_offset + ptr);
            for (tag, value) in &exif_tags {
                if let Some(name) = tag_name(*tag) {
                    result.insert(name.to_string(), value.clone());
                }
            }
        }

        // Parse GPS IFD
        if let Some(ptr) = gps_ifd_ptr {
            let gps_tags = self.parse_ifd(self.tiff_offset + ptr);
            let mut gps_data: HashMap<String, serde_json::Value> = HashMap::new();
            for (tag, value) in &gps_tags {
                if let Some(name) = gps_tag_name(*tag) {
                    gps_data.insert(name.to_string(), value.clone());
                }
            }

            // Convert GPS to decimal degrees
            if let (Some(lat_arr), Some(lon_arr)) = (gps_data.get("GPSLatitude"), gps_data.get("GPSLongitude")) {
                let lat_ref = gps_data.get("GPSLatitudeRef")
                    .and_then(|v| v.as_str()).unwrap_or("N");
                let lon_ref = gps_data.get("GPSLongitudeRef")
                    .and_then(|v| v.as_str()).unwrap_or("E");

                if let (Some(lat_vals), Some(lon_vals)) = (lat_arr.as_array(), lon_arr.as_array()) {
                    let lat = gps_to_decimal(lat_vals, lat_ref);
                    let lon = gps_to_decimal(lon_vals, lon_ref);
                    result.insert("GPSLatitude".to_string(), serde_json::json!(lat));
                    result.insert("GPSLongitude".to_string(), serde_json::json!(lon));
                }
                if let Some(alt) = gps_data.get("GPSAltitude") {
                    result.insert("GPSAltitude".to_string(), alt.clone());
                }
            }
        }

        result
    }

    fn parse_ifd(&self, offset: usize) -> Vec<(u16, serde_json::Value)> {
        let mut tags = Vec::new();
        if offset + 2 > self.data.len() { return tags; }

        let entry_count = self.read_u16(offset) as usize;
        let mut pos = offset + 2;

        for _ in 0..entry_count {
            if pos + 12 > self.data.len() { break; }

            let tag = self.read_u16(pos);
            let data_type = self.read_u16(pos + 2);
            let count = self.read_u32(pos + 4) as usize;
            let value_offset = pos + 8;

            let value = self.read_tag_value(data_type, count, value_offset);
            tags.push((tag, value));
            pos += 12;
        }
        tags
    }

    fn read_tag_value(&self, data_type: u16, count: usize, value_offset: usize) -> serde_json::Value {
        let type_size = match data_type { 1 | 2 => 1, 3 => 2, 4 => 4, 5 => 8, _ => 1 };
        let total_bytes = type_size * count;
        let data_offset = if total_bytes > 4 {
            self.tiff_offset + self.read_u32(value_offset) as usize
        } else {
            value_offset
        };

        match data_type {
            1 => { // BYTE
                if count == 1 { serde_json::json!(self.data.get(data_offset).copied().unwrap_or(0)) }
                else { serde_json::json!(self.data[data_offset..data_offset.min(self.data.len())].iter().take(count).collect::<Vec<_>>()) }
            }
            2 => serde_json::json!(self.read_string(data_offset, count)), // ASCII
            3 => { // SHORT
                if count == 1 { serde_json::json!(self.read_u16(data_offset)) }
                else { serde_json::json!((0..count).map(|i| self.read_u16(data_offset + i * 2)).collect::<Vec<_>>()) }
            }
            4 => { // LONG
                if count == 1 { serde_json::json!(self.read_u32(data_offset)) }
                else { serde_json::json!((0..count).map(|i| self.read_u32(data_offset + i * 4)).collect::<Vec<_>>()) }
            }
            5 => { // RATIONAL
                if count == 1 { serde_json::json!(self.read_rational(data_offset)) }
                else { serde_json::json!((0..count).map(|i| self.read_rational(data_offset + i * 8)).collect::<Vec<f64>>()) }
            }
            _ => serde_json::json!(self.read_u32(value_offset)),
        }
    }
}

fn gps_to_decimal(dms: &[serde_json::Value], reference: &str) -> f64 {
    if dms.len() < 3 { return 0.0; }
    let d = dms[0].as_f64().unwrap_or(0.0);
    let m = dms[1].as_f64().unwrap_or(0.0);
    let s = dms[2].as_f64().unwrap_or(0.0);
    let mut decimal = d + m / 60.0 + s / 3600.0;
    if reference == "S" || reference == "W" { decimal = -decimal; }
    (decimal * 1_000_000.0).round() / 1_000_000.0
}

pub struct ExifExtractEnricherProvider;

impl ExifExtractEnricherProvider {
    pub fn new() -> Self {
        Self
    }

    pub fn enrich(
        &self,
        item: &ContentItem,
        _config: &EnricherConfig,
    ) -> Result<EnrichmentResult, EnricherError> {
        let image_bytes = base64_decode_bytes(&item.content);
        let mut reader = ExifReader::new(image_bytes);
        let exif_data = reader.parse();

        let has_data = !exif_data.is_empty();
        let has_gps = exif_data.contains_key("GPSLatitude");
        let tag_count = exif_data.len();

        let camera = serde_json::json!({
            "make": exif_data.get("Make"),
            "model": exif_data.get("Model"),
            "lens": exif_data.get("LensModel"),
            "lensMake": exif_data.get("LensMake"),
        });
        let settings = serde_json::json!({
            "exposureTime": exif_data.get("ExposureTime"),
            "fNumber": exif_data.get("FNumber"),
            "iso": exif_data.get("ISOSpeedRatings"),
            "focalLength": exif_data.get("FocalLength"),
            "focalLength35mm": exif_data.get("FocalLengthIn35mmFilm"),
        });
        let datetime = serde_json::json!({
            "original": exif_data.get("DateTimeOriginal"),
            "digitized": exif_data.get("DateTimeDigitized"),
            "modified": exif_data.get("DateTime"),
        });
        let dimensions = serde_json::json!({
            "width": exif_data.get("ExifImageWidth"),
            "height": exif_data.get("ExifImageHeight"),
            "orientation": exif_data.get("Orientation"),
        });
        let gps = if has_gps {
            serde_json::json!({
                "latitude": exif_data.get("GPSLatitude"),
                "longitude": exif_data.get("GPSLongitude"),
                "altitude": exif_data.get("GPSAltitude"),
            })
        } else {
            serde_json::Value::Null
        };

        let mut fields = HashMap::new();
        fields.insert("exif".to_string(), serde_json::json!({
            "camera": camera, "settings": settings,
            "datetime": datetime, "dimensions": dimensions, "gps": gps,
        }));
        fields.insert("raw_tags".to_string(), serde_json::json!(exif_data));
        fields.insert("has_gps".to_string(), serde_json::json!(has_gps));
        fields.insert("tag_count".to_string(), serde_json::json!(tag_count));

        let mut metadata = HashMap::new();
        metadata.insert("provider".to_string(), serde_json::json!(PROVIDER_ID));
        metadata.insert("parsedTags".to_string(), serde_json::json!(tag_count));
        metadata.insert("method".to_string(), serde_json::json!("binary_ifd_parse"));

        Ok(EnrichmentResult {
            fields,
            confidence: if has_data { 0.95 } else { 0.1 },
            metadata: Some(metadata),
        })
    }

    pub fn applies_to(&self, schema: &SchemaRef) -> bool {
        let image_schemas = ["image", "photo", "jpeg", "jpg", "tiff", "picture"];
        let name_lower = schema.name.to_lowercase();
        image_schemas.iter().any(|s| name_lower.contains(s))
    }

    pub fn cost_estimate(&self, _item: &ContentItem) -> CostEstimate {
        CostEstimate { tokens: None, api_calls: Some(0), duration_ms: Some(5) }
    }
}

fn base64_decode_bytes(input: &str) -> Vec<u8> {
    let table = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    let mut output = Vec::new();
    let chars: Vec<u8> = input.bytes().filter(|b| !b.is_ascii_whitespace()).collect();
    for chunk in chars.chunks(4) {
        let mut buf = [0u8; 4];
        let mut count = 0;
        for (i, &byte) in chunk.iter().enumerate() {
            if byte == b'=' { break; }
            if let Some(pos) = table.iter().position(|&c| c == byte) {
                buf[i] = pos as u8;
                count = i + 1;
            }
        }
        if count >= 2 { output.push((buf[0] << 2) | (buf[1] >> 4)); }
        if count >= 3 { output.push((buf[1] << 4) | (buf[2] >> 2)); }
        if count >= 4 { output.push((buf[2] << 6) | buf[3]); }
    }
    output
}
